### DISCUSSION TITLE
So, how to actually use userscripts? :)

### DISCUSSION THREAD

**Discussion by 097115** [Jelmerro/Vieb#537] [Q&A] [answered]
Hi!

From the docs, I understand that Vieb supports Greasemonkey-like userscripts. So, I create some simple script, like this:

    // ==UserScript==
    // @name            Twitter to Nitter
    // @description     redirects Twitter to Nitter
    // @namespace       some
    // @author          some
    // @match           https://*.twitter.com/*
    // @grant           none
    // @run-at          document-start
    // ==/UserScript==
    if (location.href.includes("mobile.twitter.com")) {
        window.location.replace(window.location.href.replace("mobile.twitter.com", "nitter.net"));
    } else {
        window.location.replace(window.location.href.replace("twitter.com", "nitter.net"));
    }

Then, save it as `twitter-nitter.user.js` in `~/.config/Vieb/userscript/gm/` or ir `~/.vieb/userscript/gm/` (I'm on Ubuntu), add `set userscript` to `~/.viebrc`, launch Vieb (version 11.0.0), open a random Twitter link, and -- no redirect happens :) (it does happen in Firefox with Tampermonkey, though).

So, what am I missing? :)

**Comment by Jelmerro** [comment]
Okay, so there are 2 bugs that cause this to not work, I'll fix both in the future, but for now you could make it work by changing the following:

- Add a blank line between the comments above and the if statement, this causes the script to be mixed with the header at the moment, this is what is actually breaking the script currently.
- Secondly, I wasn't aware that single values were accepted for the match option, I thought those should always be arrays, so that's not actually being parsed, meaning this will run on any page currently. So for now, this script will work as expected:

```js
// ==UserScript==                                                               
// @name            Twitter to Nitter                                           
// @description     redirects Twitter to Nitter                                 
// @namespace       some                                                        
// @author          some                                                        
// @match           https://*.twitter.com/*                                     
// @grant           none                                                        
// @run-at          document-start                                              
// ==/UserScript==                                                              
                                                                                                                                           
if (location.href.includes("mobile.twitter.com")) {                             
    window.location.replace(window.location.href.replace("mobile.twitter.com", "nitter.net"));
} else if (location.href.includes("twitter.com")) {                             
    window.location.replace(window.location.href.replace("twitter.com", "nitter.net"));
}                                                                               
```

But I'll make sure that in the future your version will work only on twitter like it should. Though keep in mind that this specific userscript is not needed, as the very same can be achieved with the redirects setting, like so:
```
:set redirects+=^https?://(\w+\.)?twitter\.com/?(.*)~https://twiiit.com/$2
```

**Comment by Jelmerro** [comment]
I also noticed that the match pattern `https://*.twitter.com/*` is weird in that the `.` means that this only works for subdomains of twitter, not the main domain, so keep that in mind, the correct pattern should be `https://*twitter.com/*` instead (though it still needs to be fixed to work at all for single entries).

**Reply by 097115** [reply]
Thanks for the replies, Jelmer!

However, that still doesn't work for me :)

Apart from the blank line inserted between the header and the code, and this updated `@match` directive there shouldn't be any changes between your version and mine, is this correct?

I just get an impression my scripts don't get loaded, may be... I tried an even more basic userscript, like this:

    // ==UserScript==
    // @name            test-alert
    // @description     some
    // @namespace       some
    // @author          some
    // @version         0.1
    // @match           https://*google.com/*
    // @grant           none
    // ==/UserScript==
    
    alert("yay");

And nothing happens :)

**Reply by Jelmerro** [reply]
Seems to work here, after my local changes only on google.com, and with the current V11 release it runs on any page:
local, works on google.com:
![image](https://github.com/Jelmerro/Vieb/assets/1696674/e5f167cb-9317-4154-9873-30036cea282c)
V11 release, works everywhere:
![image](https://github.com/Jelmerro/Vieb/assets/1696674/7bce95e9-ee01-47d1-a393-7da2ee5443f5)
Can you confirm that you have succesfully enabled userscript with `:set userscript?`

**Reply by 097115** [reply]
Thank you for your support!

Here's the result of `:set userscript?`:

![](https://qu.ax/MDJH.png)

I launch Vieb, the userscripts are in `~/.config/Vieb/userscript/gm/` (just in case, `~/.config/Vieb/` folder is NOT empty, has some other subfolders, too, so Vieb seems to indeed use it) -- and nothing happens.

May be it's somehow related that I use AppImage? Can't really understand what's wrong :)

**Reply by Jelmerro** [reply]
If you open the internal devtools with `F8` or `:internaldevtools`, do you see any errors? It seems like it should be working, and it does here, even with the appimage.

**Reply by Jelmerro** [reply]
You might also want to check the regular `F12` or `:devtools` of the page, but that's only if there are errors in the script itself, which I highly doubt in this case. I pushed the fixes, so if you know how to run from source you could also try to see if that fixes it for you.

**Reply by 097115** [reply]
No errors at all in `:devtools` or `:internaldevtools` :)

I added a `console.log("hey")` call to the userscript, too, but no track of it in the console, either. Tested with 10.6.0 deb, and even tried 11.0.0 AppImage  in a Debian virtual machine. Still nothing. As if some path would be wrong, or I would have missed some intermediate step...

Will try with the next release as soon as it appears but I doubt it's Vieb's fault -- looks more like something on my side but what? :)

**Reply by Jelmerro** [reply]
Can you show me the output of `tree userscript` in a terminal when inside the `.config/Vieb` directory? I'm curious if the paths are indeed correct, it very much sounds like they are not indeed, or if userscript is disabled. Here is mine:
![image](https://github.com/Jelmerro/Vieb/assets/1696674/6c77ca36-8543-42a2-8120-6f0fe69b7498)

**Reply by 097115** [reply]
Mine is more or less the same :)

![](https://qu.ax/XeAs.png)

**Reply by Jelmerro** [reply]
And you are very sure that `userscript` is enabled? Then I don't have any clue what it could be. You did not change the userscriptscope did you? What is the output of `:set userscript? userscriptscope?` at the moment?

**Reply by 097115** [reply]
Sorry! Indeed my bad it was, something in the script itself was wrong. I copied the code once again, and now it does try to show an alert dialog on every page :)

However, the Twitter-Nitter one still doesn't redirect... You spot anything bad in [the code](https://paste.rs/V3kla)?

**Reply by 097115** [reply]
(I noticed that Vieb can make redirects itself, thank you for that hint, it's more like I'm trying to understand the rules and / or limitations of Vieb's userscripts...)

**Reply by 097115** [reply, marked as answer]
Aha, as soon as I removed Google test userscript, Twitter-Nitter one started to work... Will continue to experiment :)

Thanks once again for all your time and support !

**Reply by Jelmerro** [reply]
Glad it works now, I believe the duplicate script problem could be because of a duplicate `GM` header identifier, I'm still looking into that, hope to get it fixed soon as well.
